[ZIP TBD] Private Information Retrieval for Nullifier Exclusion Proofs (#1198)
zcash/zips: zips/draft-valargroup-nullifier-pir.md

sequential PIR queries plus the plaintext download, then appends 3
deterministic empty-subtree siblings to obtain the depth-29 authentication
path. Total bandwidth is approximately 3.5 MB on
the first query, or approximately 3.3 MB once the Tier 0 plaintext is cached
(dominated by the Tier 2 upload).

# Motivation

Several point-in-time protocols for Zcash require a proof-of-balance:
proving that specific shielded notes are unspent at a given block height.
Usecases include airdrops, stake-weighted polling, and governance voting
systems, as described in [^draft-str4d-orchard-balance-proof]. Proving your 
balance requires proving you have a note in the note commitment tree, and it
has not been spent at that height. For privacy, we do not want to directly
reveal the client's nullifier in this balance snapshot. This means the client 
must be able to create a proof that the correct nullifier for a note was not
spent at the snapshot height. To achieve this, we make a Merkle tree of all
known nullifiers at snapshot height, and in zero knowledge prove exclusion of
the user's nullifier into this snapshot height. Hence the name "Nullifier

> exclusion proof".

Reviewer: ```suggestion
the user's nullifier from the snapshot at this height — hence the name
"Nullifier exclusion proof".
```

There is a problem though, how does the user get the exclusion proof?
A user directly querying a centralized server for the exclusion proof would
reveal their nullifier to the server, breaking the privacy guarantee.
The alternative of downloading the entire set of Orchard nullifiers is 

> impractical. As of block height 3,268,870 with 49,813,784 nullifiers, it's already 1.48 GiB (assuming binary serialization). As Zcash scales, this grows unboundedly.

Reviewer: ```suggestion
impractical. As of Mainnet block height 3,268,870 with 49,813,784 nullifiers,
this would already take 1.48 GiB (assuming binary serialization). As Zcash
scales, absent fundamental changes in the protocol design, the size of this
set will grow without bound.
```

moved in a registration period". This lowers the download size to just grow in
the number of transactions during registration period. It comes at the cost of
voting friction (requiring users to move funds), safety, and anonymity as you're
not anonymous amongst all notes, only recently moved notes.

Private Information Retrieval (PIR) provides a cryptographic solution to 
retrieve the exclusion proofs. PIR allows a client to retrieve a record from a 
server-held database without the server learning which record was requested. 
The server processes the encrypted query by touching every record in the 
database, ensuring that its access pattern reveals nothing about the target. 
The client retrieves the exclusion proof from an untrusted server without 
revealing which nullifier it is checking (see [PIR Construction]).

# Privacy Implications

Query privacy rests entirely on the Regev encryption of the client's
selection vector (see [Regev Encryption]). Regev encryption ensures the query
is computationally indistinguishable from random under the LWE assumption. 
Therefore the server learns nothing about the target record. 

> Every other component - CDKS packing, modulus switching, the packing key -

Reviewer: ```suggestion
Every other component —CDKS packing, modulus switching, the packing key—
```

  the public database and processes queries statelessly. A client could
  query multiple independent servers without requiring coordination
  between them.
- Total bandwidth per query (upload plus download) under
  10MB, suitable for mobile networks.
- Exclusion-tree work stays practical on both sides of the proof:
  server-side tree construction, padding, and PIR database generation
  for each snapshot, and in-circuit
  verification of the retrieved authentication path.
- 128-bit computational security with correctness error probability at
  most $2^{-40}$. A client can detect and recover from a correctness
  error by re-issuing the query.

# Non-requirements

The following are explicitly out of scope:

- Incremental database updates. The PIR database is computed once from
  the nullifier set at a given snapshot height and is treated
  as static for the duration of the Protocol Epoch.
- Sub-second end-to-end query latency. The two sequential PIR round-trips
  impose a latency floor determined by network conditions.
- Retrieval of data other than nullifier exclusion proofs.
- PIR transport-level wire format between client and server.

> - Noise analysis. Refer to YPIR paper for noise and security analysis. We directly use the suggested values with no amendments [^YPIR].

Author: The initial security analysis draft and an accompanying script has been added here:
[7512c35](https://github.com/zcash/zips/pull/1198/commits/7512c35087d85b9b1b2abd96e0bae531685137eb)

- The outer transport and wire-level encoding of queries and responses is out of scope for this ZIP.

# High-level summary

This subsection is non-normative.

<Rest of content providing context for understanding the Specification.>

The nullifier exclusion tree is split into three tiers:

- Tier 0 contains the top levels of the tree and is downloaded in plaintext by all clients.

> - Tier 1 contains depth-11 to depth-18 subtrees and is served as a PIR database.

Reviewer: Depth 11 can be represented in $2^{11} \times 32$ bytes = 64 KiB, correct? I'll work it out in more detail, but intuitively I am expecting that it will probably be beneficial to have a larger cap (tier 0).

Btw I do not think it is clear what "depth-11 to depth-18 subtrees" means. An explanation of that should go in the context above.

Reviewer: > Subtly, the client must hash the whole Tier 0 if the protocol is to be securely client-verifiable. It cannot hash only the parts it needs, because then it could be giving an adversarial server an error oracle for which parts it needed, which would violate the privacy goals.

As discussed on an R&D call, the original reason for not wanting the client to have to re-hash the tree was that the hash function is Poseidon, which is designed to minimize circuit cost at the expense of being slow in software. This concern can be addressed while still keeping the Tier 0 download client-verifiable, by having the client hash this download with an algorithm that is fast in software (e.g. BLAKE3), and also publishing that hash as part of the poll parameters. Then, it will be publically verifiable whether the Merkle root and intermediate nodes are correctly computed in the download.

This implies that it does make sense to omit the intermediate hashes at layers just above the leaves in order to reduce the size of the Tier 0 download. For example, if Tier 0 has $2^k$ leaves but the intermediate hashes for layers $k-1$ and $k-2$ are omitted, then the Tier 0 download size will be reduced to $\sim 2.25 \times 32 \times 2^k$ rather than $\sim 3 \times 32 \times 2^k$ (e.g. from 192 KiB to 144 KiB when $k = 11$), but the client will only have to do 3 Poseidon hashes to compute any hash at layer $k-2$.

Reviewer: If we increase Tier 0 and decrease Tier 2, Tier 2 will index into level 19 with 524288 rows, up from 262144 previously.

No, not if we also reduce the effective total tree depth by the optimization I described in https://github.com/zcash/zips/pull/1198/changes#r2972496214 . For example, in this case:

> Or we can use a depth-13 Tier 0 with 32 nullifiers in each leaf, and only have a single depth-8 PIR Tier which would be much simpler (and avoid the potential error oracle attack from needing two queries).

the only remaining PIR Tier will index into layer 13 and will be no deeper than the current Tier 2. This is a strict improvement in the demands on the PIR scheme: Tier 0 is 576 KiB (assuming internal hash layers 11 and 12 are omitted as above), which is less than the previous Tier 1 upload+download but only downloaded once, Tier 1 goes away entirely (eliminating the error oracle attack), and Tier 2 is the same depth but shifted upward (each entry is $2^5$ hashes but there are $2^5$ fewer entries). Shifting it upward reduces its cost as you argued above.

The trade-off is a more expensive circuit because we give 32 nullifiers in the witness and prove that the actual nullifier is not any of them. But this is significantly less than a doubling of the circuit size (the Merkle path validity check requires 32 Sinsemilla hashes and Sinsemilla is more expensive in a circuit than Poseidon — assuming the latter is optimized for the case of more Poseidon hashes than in the Orchard circuit).

Reviewer 2: Our merkle path validity is also Poseidon in this scheme, so this should make cost net more expensive I believe?

Puncturing is awesome, and will work (at 32 punctures, or even like 2 punctures). But not clear to me that any puncturing is really blocking the NU7 vote. I'm down to get a small puncturing to the degree its still a circuit win in.

The error oracle attack in context is mitigated with us always doing the second query in the codebase as well.
Longer term, if we align on scaling Zcash, we will be using a multi-query technique, and just always doing the second query.

Reviewer: > Our merkle path validity is also Poseidon in this scheme, so this should make cost net more expensive I believe?

Oh, I didn't realize you were re-hashing the commitment tree. That's pretty expensive, although I guess it's a one-time cost (and we had anticipated doing it for the post-quantum transition).

In any case, hashing $k+1$ field elements for relatively small $k$ using Poseidon in a circuit is still very cheap, even if it's a larger proportion of the circuit cost. Replacing Sinsemilla with Poseidon is going to end up with a *much* cheaper circuit per note than the Orchard action circuit. Users don't usually have very many notes, so I don't see proving cost as likely to be much of an obstacle.

> The error oracle attack in context is mitigated with us always doing the second query in the codebase as well.

I really think that the two queries adds a lot of complexity and analysis cost to the scheme, which is why I favoured a single query. Cryptographic engineering experience suggests that it is hard to completely avoid leakage from error oracles.

Your analysis of the cost of the PIR queries seems to assume that there is not much benefit in reducing the number of leaves while increasing the leaf size. On the contrary: $k = 2$ reduces the db size by 1/4, $k = 4$ reduces it by 3/8, and $k = 8$ reduces it by 7/16. Granted, $k = 32$ was probably too much.

I saw that you're experimenting with punctured ranges [here](https://github.com/valargroup/vote-nullifier-pir/commit/eab73c997fbce497b42af598071c65ca4dc953c8). I agree that the $k = 2$ puncturing is essentially costless, but $k = 4$ for example is also extremely attractive. The leaves in that case consist of 5 field elements, which can be compressed to 1 field element using two width-3 instances of Poseidon in [Jive mode](https://hal.science/hal-04276646/document). The cost of this is very small compared to the rest of the circuit.

Also, I'm not sure why you're so focused on minimizing client hashes outside the circuit. For most client platforms, the largest client cost at these parameters will be waiting for the upload and download. The latency of hashing can be entirely hidden by the network latency, for a much larger number of client hashes than one. (You still have to pay in power consumption but that's ok.) It's very likely to be beneficial to omit the top few internal layers —at least two layers, which only requires the client to do four width-3 and one width-2 Poseidon compressions in Jive mode for $k = 4$— in order to reduce the download and database sizes. You have to hash when doing the witness computation for the circuit, anyway, and that's in practice more expensive.

Note that there's a trick that allows you to do any number of not-zero (or not-equal) tests in a circuit while only computing one inverse. Just multiply all of the terms that should not be zero together, and the result will be non-zero iff all of those terms are.

Author: > I really think that the two queries adds a lot of complexity and analysis cost to the scheme, which is why I favoured a single query.

I prototyped the single-tier PIR design in [#318](https://github.com/valargroup/Shielded-Vote/pull/318), but the resulting YPIR shape does not appear practical.

With PIR starting at height 13, the database shape becomes 8192 rows of roughly 270 KB each. In practice, this “few very wide rows” regime performs poorly for YPIR: precomputation becomes substantially heavier, online server time increases, and the response size grows significantly.

The intuition is that fewer but larger rows reduce the row dimension but increase the per-row payload, which in YPIR shifts cost into packing/compression, increases the number of packed outputs, and inflates the encrypted response size.

On the analysis side, we have already completed the YPIR audit, including the two-tier attack surface. We mitigate the previously identified cross-query oracle by catching Tier 1 decode errors/panics and still issuing the Tier 2 query, so the server does not learn whether Tier 1 decoding succeeded from the presence or absence of the second query.

As we continue working toward scale, some form of PIR tiering is likely to remain necessary.

Continuing with the single-tier design would move us away from a design whose complexity has already been analyzed, while introducing additional operational constraints that still require further investigation and benchmarking, and while increasing preprocessing and query time. So far, we have not observed a convincing end-to-end bandwidth improvement in the prototype.

More generally, future use cases are likely to push us back toward some tiered PIR design anyway.

As a result, I would propose keeping the two-tier PIR (3 total) design for now.

>  I saw that you're experimenting with punctured ranges. I agree that the k = 2 uncturing is essentially costless, but k = 4 for example is also extremely attractive

 I have reimplemented K=2 puncturing per earlier recommendations and will update the ZIP accordingly.

I agree that K=2 is essentially costless. I also explored larger values of K, but the marginal storage savings diminished as K increased, while circuit complexity rose, so I settled on K=2.

> Also, I'm not sure why you're so focused on minimizing client hashes outside the circuit.

I agree that minimizing client-side Poseidon outside the circuit is not the main objective. In fact, the current implementation already omits all internal Merkle nodes from the PIR row and rebuilds the subtree client-side.

Reviewer 2: Pretty happy with k=2 or k=4 as well.

The choice of Poseidon is done indeed to trade-off a one time hashing from all validators during vote setup, to give all clients much faster on device proving latencies.

> I really think that the two queries adds a lot of complexity and analysis cost to the scheme, which is why I favoured a single query. Cryptographic engineering experience suggests that it is hard to completely avoid leakage from error oracles.

Agreed that removing error oracles is nice, but I don't changing the current code to temporarily remove it, knowing we'll add it back makes that much sense. I'm in the camp of we will long-term be two tier PIR'ing anyway. I want us to be scaling zcash so much and growing demand, s.t. our number of nullifiers is 10x'ing each year for the next two years! Such a result would imply we cannot long-term stay in the one query world under YPIR.

We can make the error oracle solved with always doing the second query, and will need to be doing this if we did indeed grow Orchard. The vote is an ad-hoc procedure for now.

- Tier 2 contains depth-18 to depth-26 subtrees and is served as a PIR database.

Each proof retrieval consists of one plaintext download plus two
sequential PIR queries:

1. The client obtains Tier 0 and uses the target nullifier to identify
   the Tier 1 subtree index.
2. The client issues a PIR query for the corresponding Tier 1 row.
3. From the Tier 1 row, the client derives the Tier 2 row index and
   issues a second PIR query.
4. From Tier 0 and the recovered Tier 1 and Tier 2 rows, the client
   reconstructs the depth-26 authentication path used for nullifier
   non-membership.
5. The client appends the 3 deterministic empty-subtree siblings needed

>    by the depth-29 Claim circuit.

Author: > There isn't sufficient context to understand this without reading the "Tree Depth vs. Circuit Depth" section in the Rationale. It shouldn't be necessary to read Rationale to understand the Specification, and so this needs to be explained (briefly) in the high-level context above.

Clarified here: https://github.com/zcash/zips/pull/1198/changes/4dfe8d0b99e97ba20733781db3ac5df42367c6ad

> represent the "punctured" range

Great idea. This was implemented. The spec update is TBD.

> you don't need the circuit to be a fixed size to allow for future expansion to a larger nullifier set.

When we modify circuits, the verifying keys need to be updated on the vote chain. We aimed to avoid operational overhead with the padding.

Will evaluate [the proposal to remove Tier 1](https://github.com/zcash/zips/pull/1198#discussion_r3006956703) further and get back with results

UPDATE: https://github.com/zcash/zips/pull/1198#discussion_r3013538592

>    by the depth-29 Claim circuit.

Author: Going to resolve this thread since punctured ranges have been implemented. Please let me know if any concerns are outstanding.

with $q$-scale coefficients by a bounded number of multiplications by
small digits. As a result, the noise grows by roughly
$B_{\mathsf{ks}} \cdot \log_{B_{\mathsf{ks}}} q$ rather than by $q$,
which is the difference between a usable ciphertext and one that is
destroyed by noise.

This control of per-step noise growth is especially important for the
packing algorithm, which chains $\log_2 N$ automorphisms and associated
key switches in sequence. Because each step adds noise, the gadget base
$B_{\mathsf{ks}}$ is a tuning parameter: a smaller base reduces noise
growth per step, while increasing the number of decomposition components
and therefore the size of the key-switch material and the amount of
computation.

## Construction Choice

SimplePIR is the baseline design that requires a client-download hint.
Next-generation designs such as YPIR and InsPIRe aim to eliminate that
step and reduce communication.

### Why SimplePIR Is Not Enough

In SimplePIR [^SimplePIR], the database is reshaped into a
$\sqrt{N} \times \sqrt{N}$ matrix $D$ where each cell holds one byte.
To retrieve row $i$, the client constructs a unit selection vector

> $\mu$, encrypts it under Regev [^Regev05], and sends the ciphertext to

Reviewer: ```suggestion
$\mu$, encrypts it under Regev [^Regev2024], and sends the ciphertext to
```

  a single client.

Because each note's retrieval consists of two sequential PIR queries, if a
client failure occurs during the processing of the first query, the client
is still required to complete the second query, as specified in
[Query Completion Requirement]. Suppressing the second query would give
the server an error-based oracle that can leak a bit of the queried row index.
A malicious server can craft responses that cause failures during PIR
decryption, during response decoding, or during application-level validation of 
the recovered data. See [Rationale for Query Completion Requirement] for a detailed description of the attacks.

# Requirements

- No client-side preprocessing. A mobile wallet must be able to issue
  its first query without any prior download beyond the query itself
  or client state carried over from a previous session.
- Single untrusted server with no per-client state. The server holds only
  the public database and processes queries statelessly. A client could
  query multiple independent servers without requiring coordination
  between them.
- Total bandwidth per query (upload plus download) under

>   10MB, suitable for mobile networks.

Author: Addressed

[3364568](https://github.com/zcash/zips/pull/1198/commits/3364568acd812e2a8e988e8362a5b80ae8bc77f4)

  A client proves inclusion of a value in one of the tree's intervals.

# Abstract

This document specifies a private information retrieval (PIR) scheme for
privately retrieving nullifier exclusion proofs from a Zcash nullifier
set. Any point-in-time protocol requiring proof-of-balance — such as
airdrops, stake-weighted polling, or governance voting — needs to verify
that notes are unspent by checking their nullifiers against the spent
set, without revealing which nullifier is being checked.

The construction uses YPIR+SP [^YPIR], a single-server PIR protocol built on
SimplePIR [^SimplePIR] whose security depends on LWE and RLWE. YPIR+SP requires
a single untrusted server, no client-side database hint and no DB 
pre-processing per client. This makes it suited for privacy in the Zcash 
setting.

The nullifier exclusion tree is a binary merkle tree, organized into a
three-tier data structure spanning 26 levels of depth: a plaintext broadcast
tier (192 KB, cacheable), a small PIR tier (24 MB), and a large PIR tier (6 GB).
A complete authentication path is retrieved in two sequential PIR queries plus
the plaintext download, for a total bandwidth of approximately 3.3 MB per

> query (dominated by the Tier 2 upload).

Reviewer 3: The quoted numbers are presumably based on the current size of the nullifier set for... the Orchard pool? But the abstract is not stating this. Either remove the numbers that are not constant-size (and grow with the nullifier set), or reword this to move the quoted numbers into a separate sentence (or paragraph if necessary) saying "For the Orchard nullifier set of size X (as of block height Y), ..." (to make it clear that any other numbers are constant for this instantiation).

# Motivation

Several point-in-time protocols for Zcash require a proof-of-balance:
proving that specific shielded notes are unspent at a given block height.
Usecases include airdrops, stake-weighted polling, and governance voting

> systems, as described in [^draft-str4d-orchard-balance-proof]. Proving your 

Reviewer 3: To be decided by @daira.

balance requires proving you have a note in the note commitment tree, and it
has not been spent at that height. For privacy, we do not want to directly
reveal the client's nullifier in this balance snapshot. This means the client 
must be able to create a proof that the correct nullifier for a note was not

> spent at the snapshot height. To achieve this, we make a merkle tree of all

Reviewer 3: ```suggestion
spent at the snapshot height. To achieve this, we make a Merkle tree of all
```

# Abstract

This document specifies two retrieval schemes for nullifier exclusion
proofs from a Zcash nullifier set: full download and PIR. Any
point-in-time protocol requiring proof-of-balance (such as airdrops,
stake-weighted polling, or governance voting) needs to verify that notes
are unspent by checking their nullifiers against the spent set, without
revealing which nullifier is being checked.

Both schemes operate on the same nullifier exclusion tree. They differ
only in how the client retrieves authentication path data from the
server.

**Full download:** The client downloads the complete
exclusion tree data (the Tier 0 plaintext, all Tier 1 rows, and all

> leaf records) and locally computes Tier 2 internal node hashes on

Reviewer: This still ties the "full download" option to the choices of Tier 1 and Tier 2 depths, which I think are unlikely to remain as currently specified.

Reviewer: For the parameters given at https://github.com/zcash/zips/pull/1198/changes/4dfe8d0b99e97ba20733781db3ac5df42367c6ad..2a9802bdafb24149f7a49b0072992b4908a38777#r3006956703, the download need not be any more than 2.0625 GiB ($32 \times (2^{26} + 2^{21})$ bytes) even if all of the intermediate nodes up to depth 20 inclusive are given. Actually it only has to include the number of depth-21 leaves corresponding to the real number of nullifiers (32 at each leaf), so that's ~1.52 GiB ($32 \times (50000000 + 2^{21})$ bytes) for 50 million nullifiers.

Author: Replied to the tier amendment proposal here:
https://github.com/zcash/zips/pull/1198#discussion_r3013538592

This thread is contingent on that decision, so I am keeping it open for now

nullifiers.

To verify that a target nullifier $t$ falls within the exclusion range,
the circuit MUST check:

$$\mathsf{int}(t - low \bmod q_\mathbb{P}) < \mathsf{int}(width)$$

where $\mathsf{int}(\cdot)$ denotes the canonical integer representation
in $\{0, \ldots, q_\mathbb{P} - 1\}$. The subtraction is performed in
$\mathbb{F}_ {q_ \mathbb{P}}$; the comparison is an unsigned integer
comparison on the canonical representatives.

The tree builder MUST ensure that $\mathsf{int}(low) + \mathsf{int}(width) < q_\mathbb{P}$
for every leaf, so that the exclusion range does not wrap around the
field modulus. Under this invariant, $t - low$ does not wrap for any
$t$ in the range, and the single unsigned comparison is sufficient.

### Authentication Path

The PIR tiers provide 26 sibling hashes, one per tree depth from the
leaf (depth 26) to the root of the depth-26 PIR tree (depth 0). To form
the authentication path consumed by the Claim circuit, the client MUST
append 3 additional sibling hashes corresponding to the canonical empty
subtrees above that depth-26 root, yielding a complete depth-29

> authentication path. See [Tree Depth vs. Circuit Depth]. After decrypting the PIR responses for Tiers 1 and

Reviewer: I don't understand why it's necessary or useful to pad to depth 29.

Author: The motivation was avoiding chain upgrades to update verifying keys as the nullifier set grows.

Depth 29 was a conservative choice. 27 should work fine as well, allowing up to 134M nullifiers (51M today). 

The circuit cost with 29 was negligible, so we kept more buffer.

I'm currently investigating the punctured ranges with K=32 and removing one Tier so the depth 26 will likely change.

Will propose new depths that will have a tighter buffer.

Author: The motivation was avoiding chain upgrades to update verifying keys as the nullifier set grows.

Depth 29 was a conservative choice. 27 should work fine as well, allowing up to 134M nullifiers (51M today). 

I've since then implemented the "punctured ranges".  I am going to update to 1 level of padding, resulting in a height of 26

Author: Addressed

[e82f833](https://github.com/zcash/zips/pull/1198/commits/e82f83344196382bdee19fdb1517e5425f34118a)

which nullifier it is checking, the server learns nothing about the
client's query target.

### Server Data

A server supporting full download MUST make the following data available:

1. **Tier 0 data** (196,576 bytes): identical to the Tier 0 plaintext
   payload specified in [Tier 0: Plaintext Broadcast (Depths 0–11)].

2. **All Tier 1 rows** ($2{,}048$ rows, each 12,224 bytes = 25,034,752
   bytes total): the complete set of Tier 1 row data as specified in
   [Tier 1: Small PIR (Depths 11–18)], concatenated in row-index order.

3. **All leaf records** ($n$ records, each 64 bytes): every real
   (non-empty) exclusion-range leaf in the depth-26 tree, serialized in
   ascending order by $\mathsf{low}$. Each record is a 64-byte pair
   $[\mathsf{low} \| \mathsf{width}]$ using the same encoding specified
   in [Leaf Encoding]. The server MUST also provide the count of real
   leaf records $n$ so that clients can distinguish real leaves from the
   canonical empty padding.

For the Orchard nullifier set of size 49,813,801 (as of Mainnet block
height 3,268,870), the total download is approximately 3.2 GB
(dominated by the leaf records). With standard HTTP compression

> (e.g., gzip or zstd), this is expected to be approximately 2 GB.

Reviewer: Huh? It's almost incompressible. (Strictly speaking, the nullifiers represent curve points which would allow you to compress by 1 bit, and they are sorted which would allow using delta compression by ~20 bits per field element, but I don't think it's worth the complexity.)

However, as I said [here](https://github.com/zcash/zips/pull/1198/changes/4dfe8d0b99e97ba20733781db3ac5df42367c6ad..2a9802bdafb24149f7a49b0072992b4908a38777#r3012600122), there's no need to pad the number of nullifiers to a power of 2. It should take ~1.52 GiB.

Author: Addressed [2ceb1be](https://github.com/zcash/zips/pull/1198/commits/2ceb1be4c55b3418c3b8647f052621f4c2b84e26)
